[BE] Add smoke test for [escn,gemnet,equiformer_v2] train+predict, Add optimization test for [escn,gemnet,equiformer_v2]

misko opened the pull request:
Add a test case for training a simple eSCN model on a small subset of OC22 (first 16 elements of train). 

To suppor this I added the pickle dataset format. This is a simple list of pytorch geometric objects returned from the lmdb dataloaders. 

```
def lmdb_dataset_to_pkl(lmdb_src, pickle_output_filename, n=128):
    ds = LmdbDataset(
        {
            "src": lmdb_src,
        }
    )
    pickle.dump([ds[idx] for idx in range(n)], open(pickle_output_filename, "wb"))
```
^ generate a pickle dataset from an existing lmdb dataset

We can use tests like this in the main repo and also in developing experimental branches to automatically test and confirm that we did not obviously break training.

mshuaibii commented on main.py:
```diff
@@ -81,12 +85,12 @@ def checkpoint(self, *args, **kwargs):
 
     else:  # Run locally
-        Runner()(config)
+        Runner(args)(config)
```
What's the reasoning for including `args` in this class?

mshuaibii commented:
Documenting our chat offline:

- Introducing another dataset is unnecessary, we can leverage existing readers even for the unit tests.
- Rather than uploading the datasets directly to the repo, we can upload them to s3 and download them as part of the tests. The same way we currently handle checkpoints.
- It would be great to include eqv2 instead or alongside escn as its the more commonly used model.
- We should make sure to test the other modes as well

misko commented:
> Documenting our chat offline:
> 
> * Introducing another dataset is unnecessary, we can leverage existing readers even for the unit tests.
I removed the pickle dataset, and used @r-barnes commit to use the tutorial dataset.
> * Rather than uploading the datasets directly to the repo, we can upload them to s3 and download them as part of the tests. The same way we currently handle checkpoints.
> * It would be great to include eqv2 instead or alongside escn as its the more commonly used model.
Added tests for eqv2 and gemnet
> * We should make sure to test the other modes as well
Added test for predict mode (alongside the existing train mode test)

lbluque commented on tests/e2e/test_s2ef.py:
```diff
@@ -0,0 +1,409 @@
+    return tutorial_dataset_path / "s2ef/val_20"
+
+
+def oc20_lmdb_train_config_from_path(src):
```
looks like this is only used once in oc20_lmdb_train_and_val_from_paths below, itll be easier to read the code if this dictionary is explicitly defined in there rather than this helper function

misko replied:
definitely fixed!, some of these functions were just not properly cleaned up after a long battle with pytest. I think they look better now

lbluque commented on tests/e2e/test_s2ef.py:
```diff
@@ -0,0 +1,409 @@
+    }
+
+
+def oc20_lmdb_val_config_from_path(src):
```
Same as the comment above

lbluque commented on tests/e2e/test_s2ef.py:
```diff
@@ -0,0 +1,409 @@
+    return {"src": src}
+
+
+def oc20_lmdb_train_and_val_from_paths(train_src, val_src, test_src=None):
```
Assuming this is a function so that it is easy to add additional train, val, test src to the tests.

Would it be better to make this a fixture (and if needed can it can be parametrized with different dataset src when the time comes)

misko replied:
That was the intention, its only used in two ways in the tests and never both ways back to back. 
One way is for training (train=tutorial_dataset,val=tutorial_dataset)
And the other in testing (train=tutorial_dataset,val=tutorial_dataset,test=tutorial_dataset)
I don't know how to parameterize it in some way to make it more readable. The examples i see online seem to show when we want to use all parameterizations of the fixture back to back.

misko commented on tests/e2e/test_s2ef.py:
```diff
@@ -0,0 +1,409 @@
+            "distributed_backend": "nccl",
+            "no_ddp": False,
+        }
+    )
```
fixed!

misko commented on tests/e2e/test_s2ef.py:
```diff
@@ -0,0 +1,409 @@
+        [
+            pytest.param("gemnet", 0.4, 0.06, id="gemnet"),
+            pytest.param("escn", 0.4, 0.06, id="escn"),
+            pytest.param("equiformer_v2", 0.4, 0.06, id="equiformer_v2"),
```
Training is not reproducible to a reasonable accuracy :'( just due to numerical instability. In the above we want the metrics to be less than what is being specified not more. I dont think syrupy has a good way of doing this, but maybe i missed it

lbluque reviewed (approved):
lgtm - thanks @misko!
